File: Observer/Ruby/AverageObserver.rb
require_relative "IObserver"

class AverageObserver < IObserver
    def initialize
        @averageCPU = 0.0
        @averageRAM = 0.0
        @counter = 0
        @resources = []
    end

    def update(cpu, ram)
        @resources << [cpu, ram]
        tempCPU = 0
        tempRAM = 0
        @counter += 1

        @resources.each do |resource|
            tempCPU += resource[0]
            tempRAM += resource[1]
        end

        @averageCPU = tempCPU.to_f / @counter
        @averageRAM = tempRAM.to_f / @counter

        puts "Current Average CPU: #{@averageCPU}\nCurrent Average RAM: #{@averageRAM}"
    end
end

File: Observer/Javascript/AverageObserver.js
import IObserver from "./IObserver.js";

class AverageObserver extends IObserver {
  constructor() {
    super();
    this.averageCPU = 0.0;
    this.averageRAM = 0.0;
    this.counter = 0;
    this.resources = [];
  }

  update(cpu, ram) {
    this.resources.push([cpu, ram]);
    let tempCPU = 0.0;
    let tempRAM = 0.0;
    this.counter++;

    this.resources.forEach(resource => {
      tempCPU += resource[0];
      tempRAM += resource[1];
    });

    this.averageCPU = tempCPU / this.counter;
    this.averageRAM = tempRAM / this.counter;

    console.log(`Current Average CPU: ${this.averageCPU}\nCurrent Average RAM: ${this.averageRAM}`);
  }
}

export default AverageObserver;